doc/api/zlib.md

```diff
@@ -4,60 +4,128 @@
+
+pipeline(source, gzip, destination, (err) => {
+  if (err) {
+    console.error('An error occurred:', err);
+    process.exit(1);
+  }
+});
+
+// Or, Promisified
+
+const { promisify } = require('util');
+const pipe = promisify(pipeline);
+
+async function do_gzip(input, output) {
+  const gzip = createGzip();
+  const source = createReadStream(input);
+  const destination = createWriteStream(output);
+  await pipe(source, gzip, destination);
+}
+
+do_gzip('input.txt', 'input.txt.gz')
+  .catch((err) => {
+    console.error('An error occurred:', err);
+    process.exit(1);
   });
 ```
 
 It is also possible to compress or decompress data in a single step:
 
 ```js
+const { deflate, unzip } = require('zlib');
+
 const input = '.................................';
-zlib.deflate(input, (err, buffer) => {
-  if (!err) {
-    console.log(buffer.toString('base64'));
-  } else {
-    // handle error
+deflate(input, (err, buffer) => {
+  if (err) {
+    console.error('An error occurred:', err);
+    process.exit(1);
   }
+  console.log(buffer.toString('base64'));
 });
 
 const buffer = Buffer.from('eJzT0yMAAGTvBe8=', 'base64');
-zlib.unzip(buffer, (err, buffer) => {
-  if (!err) {
-    console.log(buffer.toString());
-  } else {
-    // handle error
+unzip(buffer, (err, buffer) => {
+  if (err) {
+    console.error('An error occurred:', err);
+    process.exit(1);
   }
+  console.log(buffer.toString());
 });
+
+// Or, Promisified
+
+const { promisify } = require('util');
+const do_unzip = promisify(unzip);
+
+do_unzip(buffer)
+  .then((buf) => console.log(buf.toString()))
+  .catch((err) => {
+    console.error('An error occurred:', err);
+    process.exit(1);
+  });
 ```
 
 ## Threadpool Usage
 
-All zlib APIs, except those that are explicitly synchronous, use libuv's
-threadpool. This can lead to surprising effects in some applications, such as
-subpar performance (which can be mitigated by adjusting the [pool size][])
-and/or unrecoverable and catastrophic memory fragmentation.
+All zlib APIs, except those that are explicitly synchronous, use Node.js'
+internal threadpool. This can lead to surprising effects and performance
+limitations in some applications.
+
+Specifically, creating and using a large number of zlib objects
+simultaneously can cause significant and unrecoverable memory fragmentation
+errors to occur that can be catastrophic to the Node.js process.
+
+This problem is illustrated in the following example:
+
+```js
+const zlib = require('zlib');
+
+const payload = Buffer.from('This is some data');
+
+// Caution: do not do this
+for (let i = 0; i < 30000; ++i) {
+  zlib.deflate(payload, (err, buffer) => {});
+}
+```
+
+In this example, 30,000 deflate instances are created concurrently and
+contend with each other for execution on the limited threadpool. Because of
+how the compression algorithms allocate and use memory internally, this
+contention leads to significant fragmentation of memory.
```

Comment:
This isn’t quite right – it has almost nothing to to with how compression algorithms use or allocate memory or the fact that the threadpool is limited in size (or what that size is). It’s a platform-specific issue with the underlying system allocator.

So instead how about:

```
In this example, 30,000 deflate instances are created concurrently. Because of
how some operating systems handle memory allocation and de-allocation
in this case, this may lead to to significant memory fragmentation.
```

Reply:
Please take a look now :-)